### ISSUE TITLE
Develop and Train the Prediction Model

### ISSUE THREAD

**Issue by judejudo** [openchlai/ai#49]
# 🧠 Git Task: Train Initial Prediction Model for Modular AI Pipeline

## 📘 Description

Develop a baseline **modular machine learning pipeline** to process audio recordings into structured case summaries for child protection and gender-based violence (GBV) cases. The pipeline will be composed of distinct AI modules:

1. **Transcription (ASR)**
2. **Translation (NLLB/M2M100/Argos)**
3. **Named Entity Recognition (NER)**
4. **Classification (Case Category + Interventions)**
5. **Summarization (Case Summary Generation)**

The baseline model will be trained using **labeled data from Issue #1**. The entire process should be reproducible and tracked using **DVC** or **MLflow**. Final outputs must adhere to the following JSON format:

```json
{
  "model": "mistral",
  "created_at": "2025-05-28T07:06:58.437500914Z",
  "response": {
    "case_summary": "...",
    "named_entities": {
      "persons": [],
      "locations": [],
      "dates": []
    },
    "classification": {
      "category": [],
      "interventions_needed": [],
      "priority_level": "medium"
    },
    "case_management": {
      "safety_planning": {
        "immediate_actions": [],
        "long_term_measures": []
      },
      "psychosocial_support": {
        "short_term": [],
        "long_term": []
      },
      "legal_protocols": {
        "applicable_laws": [],
        "required_documents": [],
        "authorities_to_contact": []
      },
      "medical_protocols": {
        "immediate_needs": [],
        "follow_up_care": []
      }
    },
    "risk_assessment": {
      "red_flags": [],
      "potential_barriers": [],
      "protective_factors": []
    },
    "cultural_considerations": []
  },
  "done": true,
  "done_reason": "stop"
}
```

### 🎯 Model Performance

- [ ] - Achieve >75% classification accuracy on validation
- [ ] - Record precision, recall, F1 score in metrics_v1.json

### 📦 Reproducibility

- [ ] - Training pipeline is fully reproducible with DVC or MLflow
- [ ] - All code, configs, and artifacts are versioned
- [ ] 

### 🧹 Data Integrity

- [ ] - All data preprocessing and feature engineering steps are tracked
- [ ] - Labeled dataset is stored/versioned in MinIO/S3

### 📁 Deliverables

- [ ] - train_model.py: End-to-end training script
- [ ] - model_v1.pkl: Serialized model (pickle or ONNX)
- [ ] - metrics_v1.json: Training/validation metrics
- [ ] - inference_pipeline.py: Combines all modules into a unified inference workflow
- [ ] - output_sample.json: Final output conforming to the structure above

### 🧼 Data Preprocessing & Feature Engineering

- [x] Load labeled case transcripts
- [ ] - Normalize and tokenize input text
- [ ] - Handle missing values, language codes, and speaker metadata

### 🏗️ Model Design
Choose and implement:

- [x]  Transcription: whisper, wav2vec2
- [ ] - Translation: NLLB, M2M100, Argos
- [ ] - NER: flair, spaCy, bert-ner
- [ ] - Classification: DistilBERT, RoBERTa
- [ ] - Summarization: Mistral, FLAN-T5, BART

### 🧪 Training

- [ ] - Train the baseline classification model on labeled examples
- [ ] - Fine-tune summarization using aligned case descriptions and summaries
- [ ] - Validate and tune hyperparameters

### 📊 Evaluation

- [ ] - Evaluate classification and summarization separately
- [ ] - Log metrics and visualize results

### 🧬 Modular Inference Pipeline

- [ ] - Chain each component into an integrated inference workflow
- [ ] - Ensure input-output compatibility at each step
- [ ] - Validate against test audio files

### 💾 Save Artifacts

- [ ] - Save final trained models and upload to S3
- [ ] - Generate example output JSON

### 🧾 Documentation

- [ ] - Document models used, training configs, and data flows
- [ ] - Add instructions for training + local inference
- [ ] - Create README.md in /pipeline/ directory

## 🛠️ Dependencies

- [ ] - Issue #1: Curate and Label Training Data
- [ ] - Access to MinIO or local S3 for datasets and artifacts
- [ ] - GPU runtime (preferred) for training large transformer models

## ⏳ Effort Estimate
2 weeks
Assigned to: @ml-team, @ai-lead

## 🌍 Notes
- All models must be open-source and capable of offline inference
- Ensure alignment with child protection, legal, and cultural domain constraints
- Final pipeline will be deployed and orchestrated via Celery and monitored through ELK + Prometheus

**Comment by openchlai** [comment]
- [ ] Case Category Classifier
- [ ] Needs Assessment Classifier 
- [ ] NER

**Comment by shemmiriam** [comment]
[training_metrics.json](https://github.com/user-attachments/files/20410624/training_metrics.json)

This is confirmed and uploaded.

**Comment by shemmiriam** [comment]
train_model.py script is Uploaded to the repo
